designs/2022-forbid-disable/README.md

```diff
@@ -0,0 +1,93 @@
+- Repo: eslint/eslint
+- Start Date: 2022-02-28
+- RFC PR:
+- Authors: Brian Bartels
+
+# ESLintRC Option To Enable noInlineConfig For Specific Rules
+
+## Summary
+
+Allow users to forbid users from configuring certain rules inline. This is an extension to `noInlineConfig` where the primary difference is that `noInlineConfig` currently is set for all rules.
+
+## Motivation
+
+When working with complex lint configs. Some rules are more important than other. For this reason it can be inconsequential to disable certain rules. But incredibly bad to disable others.
+
+EG:
+```
+//eslint-disable-next-line @government/required-SECURITY-rule
+```
+Is a much worse rule to disable than
+```
+//eslint-disable-next-line noExtraSemi
+```
+
+`noInlineConfig` currently works well for some. But in reality, many users disable rules at least somewhat frequently for valid reasons. And in large repos, there are going to be exceptions that need inline configuration. 
+
+## Detailed Design
+
+I will add the option to pass `noInlineConfig` an array of rule names. This is in addition to the already supported boolean setting. This will result in those rules having inline configuration disabled.
+
+This means that I will need to modify `config-schema.js`
+```
+noInlineConfig: { type: ["boolean", "array"] },
+```
+
+EG `// eslint-disable semi` will have ~~semi~~ as an error, as well as the violation that it is attempting to disable also having an error.
+
+The error on the inline config will have a severity of 2. See [LintMessage#severity](https://eslint.org/docs/developer-guide/nodejs-api#-lintmessage-type)
+
+first in `getDirectiveComments` I will match any directives that affect a rule listed in the array `noInlineConfig`. If there is a match, I will create a problem on that line to alert the user that they are attempting to disable a rule that will not be affected by the inline config.
+
+This error will **not** have a quick fix available
+
+
+In `applyDirectives` (`apply-disable-directives.js`) I will reference the problem. If that problem matches a name in the array `noInlineConfig` I will not apply any directives. This will ensure that `problems.push(problem)` gets run. Note that this same logic applies to configuration comments as well.
+
+## Documentation
+
+We will want to update the documentation ([here](https://eslint.org/docs/user-guide/configuring/rules#disabling-inline-comments)) to reflect the new changes. The announcement can be integrated into the config update announcement.
+
+## Drawbacks
```

Comment:
> When someone enables a new rule in a codebase, they often believe that their new rule should be followed 100% and that no one should ever disable it. But there are almost always edge cases to any rule. And so likely, a rule that begins in `noInlineConfig` will need to be removed eventually when reality sets in that the rule doesn't handle every edge case.
> 
> And how will users know what to do when they run into an edge case that a `noInlineConfig` rule doesn't handle? Will they know it's time to remove the rule from this list in the global config file? Previously, if they ran into an edge case, they could just write `// eslint-disable-line some-rule -- good reason for why this rule needs to be disabled here`.

This really seems like a big concern. If it happens that a `noInlineConfig` rule must be, for whatever valid reason (e.g., a false positive), disabled in at least one place in the codebase, then the whole concept fails? Ideally, someone would have a way to approve that particular disabling instance instead of removing the rule from `noInlineConfig` and thus allowing all future instances without control.

We recently implemented exporting information about suppressed problems:

https://github.com/eslint/rfcs/tree/main/designs/2021-suppression-support

Motivation, from the RFC and the [original issue](https://github.com/eslint/eslint/issues/14784):

> In our scenarios, when violations are suppressed, Security Development Lifecycle (SDL) tools, such as ESLint, are expected to export suppression information, including its kind and justification. We could take advantage of this info to generate corresponding signals for auditing purposes.

> This suppression info would trigger a signal to our team, and thus we could judge whether the suppression is reasonable by the justification.

In my understanding, in combination with some tools that probably already exist, this feature allows for a workflow where each new suppression triggers a review from a dedicated team and requires approval, which could be a better solution for the problem this RFC aims to solve.

Reply:
I'm confused how the supressions affect this? As it currently stands, users can disable all rules that they wish without triggering an error from ESLint. Without my RFC or `noInlineConfig` All rules that attempt to solve this problem (whose existence is evidence that it's a fairly common problem) can be overridden with ease

Reply:
It's additional info intended for use by tools that analyze linting results. The original use case is for requiring separate approvals for newly introduced eslint disable comments. which I thought would be helpful for the problem you want to solve, but it can be utilized in any other way.

If the goal is only to trigger a non-zero exit code when forbidden rules are disabled, you could run eslint with `-f json -o results.json` on CI, and then run a script that reads the json file and exits with 1 if it finds `eslint-comments/no-restricted-disable` in `suppressedMessages`.

Reply:
I think we may be trying too hard to come up with a technical solution to a social problem. Teams can discuss among themselves which rules to disallow inline config for, and they can iterate if their config isn’t working for their use case. I don’t want to get bogged down worry about how people will figure out how to properly configure ESLint for their team — they already have a process for that.

Reply:
After thinking on this. I believe that the original comment in this thread actually hints for a very GOOD reason that we'd want this enhancement. Since rules inevitably have exceptions. noInlineConfig as it stands now is essentially unusable for any medium or larger sized repo. If there are rules that need stricter enforcement. As it stands, there is nothing that really exists

Reply:
After catching up on this RFC and this discussion, I’ve seen a problem and a hypothesis. Problem: in large teams, existing automated (CI) and manual (code review) workflows are insufficient to prevent developers from using `eslint-disable` as an improper shortcut. Hypothesis: bringing changes that `eslint-disable` sensitive rules to the attention of code owners or required reviewers would protect against improper shortcuts.

We’ve agreed that this is a human problem. Configurability is a core design principle of ESLint, so a sufficiently determined developer with a deadline can circumvent any technological hurdle we erect. Looping in a code owner who’s familiar with the original rationale for the rule could act as a deterrent or would at least create a discussion.

With that in mind, existing tools like `eslint-comments/no-restricted-disable` are insufficient because they’re purely technological and can be disabled without drawing human attention. 

Proposals like `forbid` or `noInlineConfig` would draw human attention, but as already mentioned, the rule could not tolerate even a single false positive alert because there would be no workaround. It’s also impossible to migrate a rule by monotonically fixing existing suppressions while preventing new ones.

I’m intrigued by the idea of using suppression info, which ESLint now provides, for this use case. Without any changes to ESLint, someone could today write a CI tool that fails for any suppressions of specific rules. Changes to the CI tool itself could automatically flag code owners for review.

To allow exceptions or gradual migration, the tool could check those suppressions against a set of approved suppressions, fail for any unapproved suppressions, or flag code owners for discussion of changes to the approved list.

Reply:
While this suppression idea is a good one. And I'll investigate it for the use case of my current large repo. I wasn't trying to shoehorn a fix in just for my one use case. How can we set this up so that future people who have this problem can easily prevent disables and other inline config hacks. Is a blog post demonstrating this angle enough? Or is this a valid enough use case to make the RFC still worth considering so that it's actually a first class experience

Reply:
IMHO, having a way to disable inline configuration of a particular rule is generic enough that it has a place in the core. I think it solves a problem that can’t easily be solved right now.

As with all features, I don’t think we need to have a solution for every edge case. I would consider disabling inline rule configuration to be a power user feature. Yes, you lose the ability to make one-off exceptions, but that is a consequence of using a power user feature. Presumably the person opting-in would be aware of this limitation and accept the consequences.

I don’t think a blog post on the idea would be a bad way to go. Getting some more thoughts from the universe of ESLint users is always a good idea.